extension/packagemanagerextension/serverextension/packagemanagement.py

```diff
@@ -0,0 +1,226 @@
+import json
+import os
+import re
+
+from subprocess import Popen
+from tempfile import TemporaryFile
+
+from pkg_resources import parse_version
+from notebook.base.handlers import (
+    APIHandler,
+    json_errors,
+)
+from tornado import web, escape
+
+from .envmanager import EnvManager
+
+from os.path import expanduser
+home = expanduser("~")
+
+def relativeDir(directory):
+    if directory[0] != '/':
+        directory = '/' + directory
+    if directory[-1] != '/':
+        directory = directory + '/'
+    return home + directory
+
+
+class EnvBaseHandler(APIHandler):
+    """
+    Maintains a reference to the
+    'env_manager' which implements all of the conda functions.
+    """
+    @property
+    def env_manager(self):
+        """Return our env_manager instance"""
+        return self.settings['env_manager']
+
+# -----------------------------------------------------------------------------
+# Package Management APIs
+# -----------------------------------------------------------------------------
+
+
+class PkgHandler(EnvBaseHandler):
+
+    def clean(self, packages):
+        # no hyphens up front, please
+        _pkg_regex = r"(?P<pkg>[^\-][\-\da-zA-Z\._]+)"
+        # don't allow arbitrary switches
+        packages = [pkg for pkg in packages if re.match(_pkg_regex, pkg)]
+        return packages
+
+    """
+    Handler for `POST /packages` which installs all
+    the selected packages in the specified environment.
+    """
+
+    @json_errors
+    def post(self):
+        data = escape.json_decode(self.request.body)
+        directory = data.get('dir')
+        directory = relativeDir(directory)
+        packages = data.get('packages')
+        packages = self.clean(packages)
+        resp = self.env_manager.install_packages(directory, packages)
+        if resp.get("error"):
+            self.set_status(500)
```

Comment:
I do not agree with returning internal server error when the packages do not get installed. Unless something unexpected happens.

Reply:
And, as I said, I would create our exceptions (or see if any of the standard ones apply) and deal with them here.